tsdb/chunks/head_chunks.go

```diff
@@ -0,0 +1,767 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package chunks
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/binary"
+	"hash"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/tsdb/chunkenc"
+	tsdb_errors "github.com/prometheus/prometheus/tsdb/errors"
+	"github.com/prometheus/prometheus/tsdb/fileutil"
+)
+
+// Head chunk file header fields constants.
+const (
+	// MagicHeadChunks is 4 bytes at the beginning of a head chunk file.
+	MagicHeadChunks = 0x0130BC91
+
+	headChunksFormatV1 = 1
+	writeBufferSize    = 4 * 1024 * 1024 // 4 MiB.
+)
+
+var (
+	// ErrChunkDiskMapperClosed returned by any method indicates
+	// that the ChunkDiskMapper was closed.
+	ErrChunkDiskMapperClosed = errors.New("ChunkDiskMapper closed")
+)
+
+const (
+	// DefaultHeadChunkFileMaxTimeRange is the default head chunk file time range.
+	// Assuming a general scrape interval of 15s, a chunk with 120 samples would
+	// be cut every 30m, so anything <30m will cause lots of empty files. And keeping
+	// it exactly 30m also has a chance of having empty files as its near that border.
+	// Hence keeping it a little more than 30m, i.e. 40m.
+	DefaultHeadChunkFileMaxTimeRange = 40 * int64(time.Minute/time.Millisecond)
+	// MintMaxtSize is the size of the mint/maxt for head chunk file and chunks.
+	MintMaxtSize = 8
+	// SeriesRefSize is the size of series reference on disk.
+	SeriesRefSize = 8
+	// HeadChunkFileHeaderSize is the total size of the header for the head chunk file.
+	HeadChunkFileHeaderSize = SegmentHeaderSize
+	// MaxHeadChunkFileSize is the max size of a head chunk file.
+	MaxHeadChunkFileSize = 512 * 1024 * 1024 // 512 MiB.
+	// CRCSize is the size of crc32 sum on disk.
+	CRCSize = 4
+	// MaxHeadChunkMetaSize is the max size of an mmapped chunks minus the chunks data.
+	// Max because the uvarint size can be smaller.
+	MaxHeadChunkMetaSize = SeriesRefSize + 2*MintMaxtSize + ChunksFormatVersionSize + MaxChunkLengthFieldSize + CRCSize
+)
+
+// corruptionErr is an error that's returned when corruption is encountered.
+type corruptionErr struct {
+	Dir       string
+	FileIndex int
+	Err       error
+}
+
+func (e *corruptionErr) Error() string {
+	return errors.Wrapf(e.Err, "corruption in head chunk file %s", segmentFile(e.Dir, e.FileIndex)).Error()
+}
+
+// ChunkDiskMapper is for writing the Head block chunks to the disk
+// and access chunks via mmapped file.
+type ChunkDiskMapper struct {
+	/// Writer.
+	dir *os.File
+
+	curFile         *os.File // File being written to.
+	curFileSequence int      // Index of current open file being appended to.
+	curFileMint     int64    // Current file's mint. Used to check for a chunk crossing the max file time range.
+	curFileMaxt     int64    // Maxt of all chunks written to current file. This is written to the file when the file is closed.
```

Comment:
The var name is quite clear that it is for the current file maxt so I am thinkig a comment as simple as.
```suggestion
	curFileMaxt     int64    // Used for the size retention.
```

Reply:
More like for `time retention`. Updated the comments.